### PR TITLE
Support Union[NoneType, T] as input type

### LINKED ISSUE #51582: [JIT] Support Union[NoneType, T] as input type
## 🐛 Bug

While `Union` is not supported yet in TorchScript, it would be tremendously useful to have support for `Union[None, ...]`. This is because the `Optional[T]` type annotation is just an alias to `Union[None, T]`, and prior to Python 3.9, `__repr__` for `Union` does not account for this special case (in Python 3.9+, `repr(Union[None, T])` will display `Optional[T]`.

This is especially necessary for code generators like torch.fx, where Optional arguments will become Union types in the resulting GraphModule, making the GraphModule non-scriptable.

I've brought this up with the FX team (see #51457) but a better solution would be to change this within the TorchScript side to account for any kind of code generator.

## REPRO
```
class TestModule(torch.nn.Module):
    def forward(self, x: torch.Tensor, k: Optional[int]):
        return x

m = TestModule()
gm = torch.fx.symbolic_trace(m)
scripted_gm = torch.jit.script(gm)
```

## EXPECTED OUTCOME
This should work without errors.
## ACTUAL OUTCOME
I get the error `Unknown type constructor Union` because the resulting GraphModule contains Union annotations in place of the Optional type annotations.

cc @gmagogsfm

### PR DESCRIPTION
**Pull request by nikithamalgifb**
ghstack-source-id: 32db9661ce0f9441ef7061285bc24967c2808ea6
Pull Request resolved: https://github.com/pytorch/pytorch/pull/51605

Fixes #51582

Summary:
=========
In Python 3.9+ Union[T, NoneType] and Union[NoneType, T] as OptionalType.

Test:
====
python test/test_jit.py -v TestJit.test_union_to_optional

### CODE REVIEW

File: torch/csrc/jit/frontend/script_type_parser.cpp [hunk 1]
```diff
@@ -70,6 +70,30 @@ TypePtr ScriptTypeParser::subscriptToType(
...
+
+    bool first_none = first_type == NoneType::get();
+    bool second_none = second_type == NoneType::get();
+
+    if (first_none && !second_none) {
+      return OptionalType::create(second_type);
+    } else if (!first_none && second_none) {
+      return OptionalType::create(first_type);
+    } else {
+      throw ErrorReport(subscript.range()) << err;
```
Comment (gmagogsfm) [inline torch/csrc/jit/frontend/script_type_parser.cpp:94]: Just to be clear, `Union[None, None]` would trigger this line and error out, is this intended?

File: test/test_jit.py [hunk 2]
```diff
@@ -2244,6 +2244,32 @@ def forward(self, input, other=four):
         t = Test()
         self.assertEqual(t(torch.ones(1)), torch.ones(1) + 4)
 
+    def test_union_to_optional(self):
+        def test1(u: Union[int, None]) -> int:
+            if u is not None:
+                return u
+            else:
+                return 0
+        scripted = torch.jit.script(test1)
```
Comment (gmagogsfm) [inline test/test_jit.py:2253]: Optional:

Use `self.checkScript`

**Review by gmagogsfm — APPROVED** [review]

File: torch/csrc/jit/frontend/script_type_parser.cpp (hunk 1 above)
Comment (nikithamalgifb) [reply torch/csrc/jit/frontend/script_type_parser.cpp:94]: Yes. That should be handled during the regular Union.

File: test/test_jit.py (hunk 2 above)
Comment (nikithamalgifb) [reply test/test_jit.py:2253]: I was told `self.checkScript` is the old way in one of my earlier PR's.

Comment (gmagogsfm) [reply test/test_jit.py:2253]: Hmm, didn't know about that. What's wrong with it and what's the alternative?

**Pull request closed without merging**